### PR TITLE
Enums for Generate Layer

### PR DESCRIPTION
**Pull request by BjarkeKarlsen**

### CODE REVIEW

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 1]
```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
...
-                // Instantiate Room Game Objects
-                InstantiateRoomGameObjects(MapBuilderRoomDictionary);
+                case Build.CreateNodeGraph:
+                    (status, rebuildAttemptsForNodeGraph) = AttemptCreating(roomNodeGraph, MapBuilderRoomDictionary, roomTemplateList, _roomNodeTypeList, rebuildAttemptsForNodeGraph);
+                    Debug.Log( $"Attempting creating: {status.ToString()}");
+                    break;
+                case Build.ReAttemptFailedCreateNodeGraph:
+                    // Check status 
+                    // If attempt failed then count up 
+                    status = (int)rebuildAttemptsForNodeGraph == Settings.MaxMapBuildAttempts
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:56]: I dont like the Settings.MaxMapBuildAttempts is used both in this and attempt failed. How should we do this? Create two parameteres or use the last enum?

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 2]
```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
...
-            }
-
-
-            if (dungeonBuildSuccessful)
-            {
-                // Instantiate Room Game Objects
-                InstantiateRoomGameObjects(MapBuilderRoomDictionary);
+                case Build.CreateNodeGraph:
+                    (status, rebuildAttemptsForNodeGraph) = AttemptCreating(roomNodeGraph, MapBuilderRoomDictionary, roomTemplateList, _roomNodeTypeList, rebuildAttemptsForNodeGraph);
+                    Debug.Log( $"Attempting creating: {status.ToString()}");
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:51]: This is just for testing: ```Debug.Log( $"Attempting creating: {status.ToString()}");```

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 3]
```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
...
+                        ? Build.AttemptFailed
+                        : Build.AttemptBuilding;
+                    break;
+                case Build.AttemptFailed:
+                    // 
+                    status = (int)buildAttempts == Settings.MaxMapBuildAttempts
+                        ? Build.MaxAttemptFailed
+                        : Build.AttemptBuilding;
+                    break;
+                case Build.AttemptBuilding:
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:66]: Should this be in a methode or is it readable enough?

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 4]
```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
...
+                case Build.MaxAttemptFailed:
+                    status = Build.Failed; 
+                    break;
+                case Build.Failed:
+                    return Build.Failed;
+                case Build.Success:
+                    InstantiateRoomGameObjects(MapBuilderRoomDictionary);
+                    return Build.Success;
             }
+            Debug.Log(status.ToString());
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:85]: Testing, but we are keeping this!

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 5]
```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
...
         }
+        
+    }
+
+    private (Build, Attempt) AttemptCreating(RoomNodeGraphSO roomNodeGraph, Dictionary<string, Room> mapBuilderRoomDictionary,
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:90]: Maybe to little happping here and is too inpure.... 

This in the switch case instead
```
ClearMap(MapBuilderRoomDictionary);
AttemptToBuildRandomMap(roomNodeGraph, MapBuilderRoomDictionary, roomTemplates, _roomNodeTypeList);
++rebuildAttemptsForNodeGraph
```

File: TheEngineeringJourney/Assets/Scripts/Enums/Enum.cs [hunk 6]
```diff
@@ -30,4 +30,37 @@ public enum AimDirection
...
+    Fourth,
+    Five,
+    Six,
+    Seven,
+    Eight,
+    Nine,
+    Ten
+}
+
+public enum RoomOverlapping
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Enums/Enum.cs:61]: Hate the name of this enum

File: TheEngineeringJourney/Assets/Scripts/GameManager/GameManager.cs [hunk 7]
```diff
@@ -147,9 +148,19 @@ private void HandleGameState()
 
     private void PlayMapLevel(int MapLevelListIndex)
     {
-        if (MapBuilder.Instance.GenerateMap(MapLevelList[MapLevelListIndex])) return;
+        var statusGenerateMap = MapBuilder.Instance.GenerateMap(MapLevelList[MapLevelListIndex]);
+        Debug.Log($"Status for generate map: {statusGenerateMap}");
         
-        Debug.LogError("Couldn't build Map from specified rooms and node graphs");
+        switch (statusGenerateMap)
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/GameManager/GameManager.cs:154]: Nice alot easier to read

File: TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs [hunk 8]
```diff
@@ -139,13 +170,13 @@ private static void ClearMap(Dictionary<string, Room> mapBuilderRoomDictionary)
...
-    private bool ProcessRoomsInOpenRoomNodeQueue(RoomNodeGraphSO roomNodeGraph, Queue<RoomNodeSO> openRoomNodeQueue, 
+    private RoomOverlapping ProcessRoomsInOpenRoomNodeQueue(RoomNodeGraphSO roomNodeGraph, Queue<RoomNodeSO> openRoomNodeQueue, 
         Dictionary<string, Room> mapBuilderRoomDictionary, 
         IReadOnlyCollection<RoomTemplateSO> roomTemplates, RoomNodeTypeListSO roomNodeTypes)
     {
-        var noRoomOverlaps = true;
+        var noRoomOverlaps = RoomOverlapping.Overlapping;
         // While room nodes in open room node queue & no room overlaps detected.
-        while (openRoomNodeQueue.Count > 0 && noRoomOverlaps)
+        while (openRoomNodeQueue.Count > 0 && noRoomOverlaps is RoomOverlapping.Overlapping)
```
Comment (BjarkeKarlsen) [inline TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs:179]: Don't like this line.... 
```
openRoomNodeQueue.Count > 0
```

**Review by DuweBTC — APPROVED** [review]

**Pull request merged**